### ISSUE TITLE
Variable types are not limited by conditional checks against None

### ISSUE THREAD

**Issue by jsoules** [microsoft/pyright#657]
For a variable foo which is typed as `Union[None, bar]`, the variable's actual type must be `bar` within any conditional branch of the form `if foo is not None:`. However, the type checker is not recognizing the limitation on the type of the variable, and is throwing up a spurious error for >, <, >=, <= operators (assuming that the variable `foo` might still be of type `None`).

Minimal example:

```
from typing import Union

timeout: Union[None, float] = None
elapsed: float = 1.0

if timeout is not None:
    if elapsed > timeout:
        print('Popup: Operator ">" not supported for types "float" and "Never"')
    elif elapsed < timeout:
        print('Popup: Operator "<" not supported for types "float" and "Never"')
    elif elapsed >= timeout:
        print('Popup: Operator ">=" not supported for types "float" and "Never"')
    elif elapsed <= timeout:
        print('Popup: Operator "<=" not supported for types "float" and "Never"')
    elif elapsed == timeout:
        print("Equality test does not show error, although timeout is still typed as Never, when it can only be float.")
```

I would expect that within the conditional check, the variable `timeout` would be recognized to be of type `float` only, since it has been checked not to be `None` in this branch.

This is running in the VS Code extension v 1.1.35 on Ubuntu 18.04. The Python version is 3.6.9.

**Comment by erictraut** [comment]
This behavior is expected. That's because Pyright employs a technique called "type narrowing" to further narrow the type of a symbol as it winds its way through the code. In your example, timeout has a type of `Union[None, float]`, but it's assigned a value of `None`. Immediately after this assignment, its type is narrowed to just `None` because that's the only possible value it can contain after the assignment. The conditional statement `if timeout is not None` further narrows the type of `timeout` to `Never`, which is Pyright's way of saying "type narrowing has eliminated all possible subtypes, so you will never get here".

Note that type narrowing relies on code flow analysis and will not cross "execution context" boundaries. (An execution context is a module, a function, or a lambda.) In your example, the assignment to `timeout` is within the same execution context as the use of `timeout`, so code flow analysis is performed and type narrowing is applied.

If we modify your example and introduce a new execution context (a function), the type checker is no longer able to narrow the type of `timeout`, so you won't see the errors in this case.

```
def foo(elapsed: float, timeout: Union[None, float]):
    if timeout is not None:
        if elapsed > timeout:
            print('Popup: Operator ">" not supported for types "float" and "Never"')
       ...
```

For more details on type narrowing, check out [this documentation](https://github.com/microsoft/pyright/blob/master/docs/internals.md#type-checking-concepts).

**Comment by jsoules** [comment]
Ah, gotcha. The example was extracted from code that wasn't originally mine and the declaration and use are a couple hundred lines apart, so I hadn't realized that it's actually unreachable code.

Thanks!
